ffi/utf8proc.lua

```diff
@@ -30,4 +30,16 @@ function Utf8Proc.lowercase(str)
     return folded_str
 end
 
+function Utf8Proc.strlen(str)
+    local str_buf = ffi.cast("const utf8proc_uint8_t *", str)
+    local buffer = ffi.new("utf8proc_int32_t[1]")
+    local count = 0
+    local pos = 0
+    while str_buf[pos] ~= 0 do
+        count = count + 1
+        pos = pos + libutf8proc.utf8proc_iterate(str_buf + pos, -1, buffer)
+    end
+    return count
+end
```

Comment:
Doesn't handle errors sanely (`utf8proc_iterate` returns a negative value on error, so moving the pointer backward may lead to "interesting" results ;)).

This should do the trick (untested):

```suggestion
function Utf8Proc.strlen(str)
    local p = ffi.cast("const utf8proc_uint8_t *", str)
    local codepoint = ffi.new("utf8proc_int32_t[1]")
    local count = 0
    local pos = 0
    local bytes = 0
    while bytes >= 0 do
        bytes = libutf8proc.utf8proc_iterate(p + pos, -1, codepoint)
        if bytes > 0 then
            count = count + 1
            pos = pos + bytes
        end
    end
    return count
end
```

Reply:
It's unclear whether it can ever return `0`, in which case, either reworking the loop to avoid the funky check by preloading an initial call or breaking on 0 inside the loop might be necessary.

Reply:
TL;DR:

```suggestion
function Utf8Proc.strlen(str)
    local p = ffi.cast("const utf8proc_uint8_t *", str)
    local codepoint = ffi.new("utf8proc_int32_t[1]")

    local bytes = libutf8proc.utf8proc_iterate(p, -1, codepoint)
    if bytes <= 0 then
        -- No valid codepoints at all?
        return 0
    end
    
    local count = 1
    local pos = bytes
    while bytes > 0 do
        bytes = libutf8proc.utf8proc_iterate(p + pos, -1, codepoint)
        if bytes > 0 then
            count = count + 1
            pos = pos + bytes
        end
    end
    return count
end
```